### PR TITLE
Avb watchdog

### PR DESCRIPTION
**Pull request by tvyas**
Initial commit to add the watchdog feature for AVB Streamhandler. 

There is an include change in the 2nd commit (build: Fix build avb_streamhandler_watchdog build), please help verify that patch and ensure that it works in someone else's build environment as well.

### CODE REVIEW

**Review by vcgomes — COMMENTED** [review]
This patch looks good, only a typo on the commit message: "inculde" should be "include".

**Review by vcgomes — APPROVED** [review]
Looks good as well.

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
+/*
+ @COPYRIGHT_TAG@
```
Comment (vcgomes) [inline private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: We are using a proper License text right now, I think this should be fixed before merging:
(Even, I think, if you aren't the original author)

```
/*
 * Copyright (C) 2018 Intel Corporation. All rights reserved.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */
```

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+      mResetTimespan = uint64_t( intValue/watchdogUSec2TimespanDivisor );
+    }
+  }
+  else
+  {
+    DLT_LOG_CXX(mDltContext, DLT_LOG_WARN, "IasSystemdWatchdogManager no watchdog configuration found in env 'WATCHDOG_USEC'. Disable checking.");
+  }
+
+  if ( result.toString() == IasMediaTransportAvb::IasResult::cOk.toString() && ( mResetTimespan > 0LL) )  //watchdog result type not compatible with IAS_SUCCEED
+  //if ( IasMediaTransportAvb::IAS_SUCCEEDED(result) && ( mResetTimespan > 0LL) )
```
Comment (vcgomes) [inline private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:151]: I noticed there are a few of these commented statements. Do they come from the original patch? If yes, I think they need to be removed in a later patch.

**Review by vcgomes — COMMENTED** [review]
It's a lot of code, so reviewing it all carefully would take a lot of time.

But I didn't find anything too scary, so I would suggest the following: first thing, change the code so it has a proper License. Then, if no one objects or has any issues, we apply this, and we work from there (for example, removing the commented out code).

What do you guys think?

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp (hunk 2 above)
Comment (tvyas) [reply private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:151]: Yes, they're all from the original patch. I believe that's true for all of AVB streamhandler code. If we get into all the cleanup that remains to be done, then it'll probably take a month to merge this. I have verified the basic functionality for this feature at this point, so, I agree that we should merge this now and going forward, we can cleanup/fix the coding style and functional issues as we come across them.

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp (hunk 1 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: True, I'll fix that up. Thanks !

Comment (tvyas) [reply private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: Done.

File: private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  : mDltContext(context)
+  , mRegistrationMutex(/*recursive_mutex*/)
+  {}
+
+  virtual ~IasWatchdogTimerRegistration() = default;
+
+  /**
+    * \brief Notifies all listeners about a new Timeout.
+    */
+  IasWatchdogResult registerTimer(IasWatchdogTimer const &timer)
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp:63]: Is someone aware that this function does...nothing?

File: private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    {
+      //result = registerTimerLocked(timer);
+    }*/
+    return result;
+  }
+
+  /**
+    * \brief Notifies all listeners about the removal of a Timeout.
+    */
+  IasWatchdogResult unregisterTimer(IasWatchdogTimer const &timer)
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp:80]: Ditto. If they should do something on the future, maybe comment all their code? Right now they appear to do something useful, while not (unless of course, some C++ incantation I'm not aware of).

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+  {
+    // Erase old timers in case of a manual call to notifyTimedOut().
+    (void)mWatchdogTimerRegistration->unregisterTimer(mWatchdogTimer);
+
+    uint64_t const nextTimeoutTimestamp = IasSystemdWatchdogManager::getCurrentRawTime() + currentResetTimespan;
+    mWatchdogTimer.setTimeoutInterval(currentResetTimespan);
+
+    IasWatchdogResult const setTimerResult = mWatchdogTimerRegistration->registerTimer(mWatchdogTimer);
+
+    if (setTimerResult.toString() == IasMediaTransportAvb::IasResult::cOk.toString())
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:81]: Why does this enum-ish thing is always compared via `toString()`? Looks funky - am I missing some C++ glory here?

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  }
+  return result;
+}
+
+IasWatchdogInterface* IasSystemdWatchdogManager::createWatchdog()
+{
+  IasWatchdogInterface * watchdogInterface = nullptr;
+  std::lock_guard<std::mutex>  mutexGuard(mWatchdogInterfacesMutex);
+
+  if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:177]: This is the same as `if (true)`, right?

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+  return watchdogInterface;
+}
+
+IasWatchdogInterface* IasSystemdWatchdogManager::createWatchdog(uint32_t timeout, std::string watchdogName)
+{
+  IasWatchdogInterface * watchdogInterface = nullptr;
+
+  std::lock_guard<std::mutex> mutexGuard(mWatchdogInterfacesMutex);
+
+  if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:199]: This is the same as `if (true)`, right?

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+{
+  if ( watchdogInterface == nullptr )
+  {
+    return IasMediaTransportAvb::IasResult::cParameterInvalid;
+  }
+  IasWatchdogResult result = IasMediaTransportAvb::IasResult::cOk;
+
+  std::lock_guard<std::mutex> mutexGuard(mWatchdogInterfacesMutex);
+
+  if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:255]: Ok, there are many of this construction...

File: private/src/avb_watchdog/IasWatchdogResult.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "avb_watchdog/IasWatchdogResult.hpp"
+#include "avb_helper/IasResult.hpp"
+
+#include <sstream>
+#include <string>
+
+#define IAS_WATCHDOG_RESULT_START_VAL 1000u
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogResult.cpp:13]: This looks fragile. Well, this class as a whole appears odd - compare with its parent, for instance.

**Review by edersondisouza — COMMENTED** [review]
Just some comments - feel free to ack or ignore them.

File: private/src/avb_watchdog/IasSystemdWatchdogManager.cpp (hunk 6 above)
Comment (tvyas) [reply private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:177]: Lol, yea. I got numb looking at this and stopped caring. But this one does warrants a modification.

File: private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp (hunk 3 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp:63]: Will remove it. If needed, we can revert back, in the future.

File: private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp (hunk 4 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasWatchdogTimerRegistration.hpp:80]: Will remove it. If needed, we can revert back, in the future.

**Comment by tvyas** [comment]
Please refer to the new pull request that addresses Ederson's comments.

**Pull request closed without merging**